File: aspnet-in-practice/08-09/CS/CoolMVBlog/Models/TagCloudService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace CoolMvcBlog.Models
{
    public class TagCloudService
    {
        private BlogModelContainer context;
        public TagCloudService(BlogModelContainer context)
        {
            this.context = context;
        }

        public List<TagCloudItem> GetTagCloudItems()
        {
            var q = (from p in context.PostSet
                    from c in p.Categories
                    group p by new { Id = c.Id, Description = c.Description } into g
                    select new
                    { 
                        CategoryId = g.Key.Id,  
                        Description = g.Key.Description,
                        Size = g.Count()
                    }).ToList();
            
            var max = q.Max(i => i.Size);

            var res = q.Select(i => new TagCloudItem()
            {
                CategoryId = i.CategoryId,
                Description = i.Description,
                Size = this.GetSize(i.Size, max)
            });

            return res.ToList();
        }

        private string GetSize(int p, int max)
        {
            var threshold1 = max * .40;
            var threshold2 = max * .80;

            if (p >= threshold2)
                return "100%";

            if (p >= threshold1)
                return "80%";

            return "60%";
        }
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Models/TagCloudService.vb
Public Class TagCloudService
    Private context As BlogModelContainer

    Public Sub New(ByVal context As BlogModelContainer)
        Me.context = context
    End Sub

    Public Function GetTagCloudItems() As List(Of TagCloudItem)
        Dim q = (From p In context.PostSet
                 From c In p.Categories
                 Group p By IdKey = c.Id, DescriptionKey = c.Description Into Group
                 Select New With
                        {
                            .CategoryId = IdKey,
                            .Description = DescriptionKey,
                            .Size = Group.Count()
                            }).ToList()

        Dim max = q.Max(Function(i) i.Size)

        Dim res = q.Select(Function(i) New TagCloudItem() With
                                       {
                                           .CategoryId = i.CategoryId,
                                           .Description = i.Description,
                                           .Size = Me.GetSize(i.Size, max)
                                       })

        Return res.ToList()
    End Function

    Private Function GetSize(ByVal size As Integer, ByVal max As Integer) As String
        Dim threshold1 = max * 0.4
        Dim threshold2 = max * 0.8

        If size >= threshold2 Then
            Return "100%"
        End If

        If size >= threshold1 Then
            Return "80%"
        End If

        Return "60%"
    End Function
End Class